iceoryx_posh/source/popo/ports/server_port_roudi.cpp

```diff
@@ -38,19 +39,108 @@ ServerPortRouDi::MemberType_t* ServerPortRouDi::getMembers() noexcept
 }
 
+ConsumerTooSlowPolicy ServerPortRouDi::getClientTooSlowPolicy() const noexcept
+{
+    return static_cast<ConsumerTooSlowPolicy>(getMembers()->m_chunkSenderData.m_subscriberTooSlowPolicy);
+}
+
 cxx::optional<capro::CaproMessage> ServerPortRouDi::tryGetCaProMessage() noexcept
 {
+    // get offer state request from user side
+    const auto offeringRequested = getMembers()->m_offeringRequested.load(std::memory_order_relaxed);
+
+    const auto isOffered = getMembers()->m_offered.load(std::memory_order_relaxed);
+
+    if (isOffered)
+    {
+        if (!offeringRequested)
+        {
+            capro::CaproMessage caproMessage(capro::CaproMessageType::STOP_OFFER, this->getCaProServiceDescription());
+            return dispatchCaProMessageAndGetPossibleResponse(caproMessage);
+        }
+    }
+    else
+    {
+        if (offeringRequested)
+        {
+            capro::CaproMessage caproMessage(capro::CaproMessageType::OFFER, this->getCaProServiceDescription());
+            return dispatchCaProMessageAndGetPossibleResponse(caproMessage);
+        }
+    }
+
     // nothing to change
-    return cxx::nullopt_t();
+    return cxx::nullopt;
 }
 
 cxx::optional<capro::CaproMessage>
-ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& /*caProMessage*/) noexcept
+ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& caProMessage) noexcept
 {
-    capro::CaproMessage responseMessage(
-        capro::CaproMessageType::NACK, this->getCaProServiceDescription(), capro::CaproMessageSubType::NOSUBTYPE);
+    const auto isOffered = getMembers()->m_offered.load(std::memory_order_relaxed);
+
+    if (isOffered)
+    {
+        return handleCapProMessageForStateOffered(caProMessage);
+    }
+    else
+    {
+        return handleCapProMessageForStateNotOffered(caProMessage);
+    }
+}
 
-    return cxx::make_optional<capro::CaproMessage>(responseMessage);
+void ServerPortRouDi::handleCaProProtocolViolation(const capro::CaproMessageType messageType) const noexcept
+{
+    // this shouldn't be reached
+    LogFatal() << "CaPro Protocol Violation! Got '" << messageType << "'!";
+    errorHandler(Error::kPOPO__CAPRO_PROTOCOL_ERROR, nullptr, ErrorLevel::SEVERE);
+}
+
+cxx::optional<capro::CaproMessage>
+ServerPortRouDi::handleCapProMessageForStateOffered(const capro::CaproMessage& caProMessage) noexcept
+{
+    capro::CaproMessage responseMessage{capro::CaproMessageType::NACK, this->getCaProServiceDescription()};
+
+    switch (caProMessage.m_type)
+    {
+    case capro::CaproMessageType::STOP_OFFER:
+        getMembers()->m_offered.store(false, std::memory_order_relaxed);
+        m_chunkSender.removeAllQueues();
+        return caProMessage;
+    case capro::CaproMessageType::OFFER:
+        return responseMessage;
+    case capro::CaproMessageType::CONNECT:
+        cxx::Expects(caProMessage.m_chunkQueueData != nullptr && "Invalid response queue passed to server");
```

Comment:
Wouldn't this statement kill roudi? So the only thing I have to do is to send an invalid connect request and everything goes down? Wouldn't it be better to send here a `NACK` and print an error into the log?

Would it make sense to also add some sanity check here and maybe verify if the `chunkQueueData` pointer is valid in the sense that it originates from a valid shared memory segment which is accessible? Otherwise some application can add a chunk queue data where the server port has no access to and then you kill the server with a segfault?